src/Moq/ProxyFactories/CastleProxyFactory.cs

```diff
@@ -136,6 +151,108 @@ public void DetachFromUnderlying()
 			}
 		}
 
+#if FEATURE_DEFAULT_INTERFACE_IMPLEMENTATIONS
+		// NOTE: In theory, this helper should work on any platform. It is excluded on `netstandard2.0`
+		// and lower mostly to avoid an additional NuGet dependency required for `DynamicMethod`.
+		private static class DefaultImplementationThunk
+		{
+			// We store previously generated dynamic thunks for reuse.
+			private static ConcurrentDictionary<MethodInfo, Func<object, object[], object>> thunks;
+
+			static DefaultImplementationThunk()
+			{
+				thunks = new ConcurrentDictionary<MethodInfo, Func<object, object[], object>>();
+			}
+
+			public static Func<object, object[], object> Get(MethodInfo method)
+			{
+				return thunks.GetOrAdd(method, static method =>
+				{
+					var originalParameterTypes = method.GetParameterTypes();
+					var n = originalParameterTypes.Count;
+
+					var dynamicMethod = new DynamicMethod(string.Empty, returnType: typeof(object), parameterTypes: new[] { typeof(object), typeof(object[]) });
+					dynamicMethod.InitLocals = true;
+					var il = dynamicMethod.GetILGenerator();
+
+					var arguments = new LocalBuilder[n];
+					var returnValue = il.DeclareLocal(typeof(object));
+
+					// Erase by-ref-ness of parameter types to get at the actual type of value.
+					// We need this because we are handed `invocation.Arguments` as an `object[]` array.
+					var parameterTypes = originalParameterTypes.ToArray();
+					for (var i = 0; i < n; ++i)
+					{
+						if (parameterTypes[i].IsByRef)
+						{
+							parameterTypes[i] = parameterTypes[i].GetElementType();
+						}
+					}
+
+					// Transfer `invocation.Arguments` into appropriately typed local variables.
+					// This involves unboxing value-typed arguments, and possibly down-casting others from `object`.
+					// The `unbox.any` instruction will do the right thing in both cases.
+					for (var i = 0; i < n; ++i)
+					{
+						arguments[i] = il.DeclareLocal(parameterTypes[i]);
+
+						il.Emit(OpCodes.Ldarg_1);
+						il.Emit(OpCodes.Ldc_I4, i);
+						il.Emit(OpCodes.Ldelem_Ref);
+						il.Emit(OpCodes.Unbox_Any, parameterTypes[i]);
+						il.Emit(OpCodes.Stloc, arguments[i]);
+					}
+
+					// Now we're going to call the actual default implementation.
+
+					// We do this inside a `try` block because we need to write back possibly modified
+					// arguments to `invocation.Arguments` even if the called method throws.
+					var returnLabel = il.DefineLabel();
+					il.BeginExceptionBlock();
+
+					// Perform the actual call.
+					il.Emit(OpCodes.Ldarg_0);
+					il.Emit(OpCodes.Castclass, method.DeclaringType);
+					for (var i = 0; i < n; ++i)
+					{
+						il.Emit(originalParameterTypes[i].IsByRef ? OpCodes.Ldloca : OpCodes.Ldloc, arguments[i]);
+					}
+					il.Emit(OpCodes.Call, method);
+
+					// Put the return value in a local variable for later retrieval.
+					if (method.ReturnType != typeof(void))
+					{
+						il.Emit(OpCodes.Box, method.ReturnType);
```

Comment:
This would fail on a by-ref return type (which AFAIK neither we nor DynamicProxy currently support anyway).